### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by jwr** [apa512/clj-rethinkdb#97]
I started working with larger data sets and I'm running out of memory. From a quick look with the yourkit profiler it seems that the heap gets filled up with `clojure.core.async$mult$reify__6387` objects, stored in a `clojure.lang.PersistentHashMap$INode[32]`.

A way to trigger this is simply to try to perform lots of operations:

```
(doseq [i (range 5000000)]
    (-> (r/db db) (r/table "docs") (r/insert [{:id i}]) (r/run *conn*)))
```

If I let the above snippet run for several minutes, memory consumption will start rising after a while, eventually exhausting the 4GB heap.

I've attached a snapshot of the memory allocation by class, hopefully this will help.
![2015-10-07 at 22 47](https://cloud.githubusercontent.com/assets/38015/10350654/8b0ac10a-6d45-11e5-97a8-719e10be0efd.png)

**Comment by danielcompton** [comment]
Does it GC at all?

**Comment by jwr** [comment]
Yes. And in the beginning things look just fine, it is only after a while that the heap starts growing. I am running 0.10.1.

I can offer a wild guess that it isn't a GC problem at all, but rather one of timing: there are leftovers of core.async structures because there is no time to process them, and things start accumulating. I suspect that if I inserted a pause every one million transactions, things might clear up (I'll actually try that).

This is a critical bug, it prevents me from using clj-rethinkdb at all.

**Comment by jwr** [comment]
Turns out my wild guess was wrong: just running that loop in 100k increments, with pauses inbetween runs, is enough to eventually exhaust the heap. Some core.async structures related to pub/sub are being retained forever.

Apparently people didn't notice it because nobody makes a lot of transactions. But every clj-rethinkdb app out there right now leaks memory.

**Comment by danielcompton** [comment]
It'd be great to know if this was a core.async issue or something that we've done wrong. My (clearly naive) assumption was that the sub channels would be GC'd when there were no references to them, but that doesn't seem to be the case.

I don't have a lot of time to look at this this week, are you able to dig into this a little bit. @danielytics may have some thoughts as he worked on this?

**Comment by jwr** [comment]
I'm sorry, there is no way I'll be able to spend time on this in the near future :-( I just have to use something else for now, I can't put the project on hold.

**Comment by danielytics** [comment]
It just occurred to me that the response `type` may be `1` for an insert, meaning that the `token` never gets removed from the `:waiting` list. Eg if `type` is `2`, this gets called https://github.com/apa512/clj-rethinkdb/blob/master/src/rethinkdb/net.clj#L116 but not if its `1` or if there's an exception. OTOH this seems like a bug and could very well be the cause of this. What do you think, @danielcompton ?

**Comment by danielcompton** [comment]
@jwr can you give me more details on your testing setup, including JVM opts, where RethinkDB is running, a graph of memory usage over time, how long the tests were running for, any logs created, what exact error you got, e.t.c.? I ran your test myself with a 512 MB (and 4GB) heap and while I could see the memory use going up, the JVM was able to successfully GC it away.

![screenshot of visualvm 9-10-15 4 21 07 pm](https://cloud.githubusercontent.com/assets/811954/10385181/e9a7426c-6ea1-11e5-908a-6921c0e8c467.png)

Over the longer term, GC's started to get more frequent, which needs to be looked into further. Still, I wasn't seeing the heap exhaustion that you described.

![screenshot of visualvm 9-10-15 4 31 19 pm](https://cloud.githubusercontent.com/assets/811954/10385321/a5dae8b6-6ea3-11e5-8c08-adb434685977.png)

**Comment by danielcompton** [comment]
This looks very suspect: [ASYNC-90](http://dev.clojure.org/jira/browse/ASYNC-90)

**Comment by danielcompton** [comment]
Running this for longer, I can see the inevitable:

![screenshot of tweetbot 9-10-15 5 37 01 pm](https://cloud.githubusercontent.com/assets/811954/10385915/664437bc-6eac-11e5-8c64-7cf537dce856.png)

**Comment by jwr** [comment]
@danielcompton I don't think the exact JVM opts are important — as you later noticed, you'll exhaust the heap eventually anyway. I normally run with default JVM settings on Mac OS X, which is equivalent to `-Xmx4096M`, I think. RethinkDB runs locally.

**Comment by danielytics** [comment]
Ouch... I believe ASYNC-90 is the cause of this. core.async creates a new mult for every topic and as ASYNC-90 says, when there are no taps left, the mult is still kept around. That means that once you sub to a topic, a mult is created that is never ever released.

We use the `token` as the topic, meaning that there is a new mult created for every single query and that never gets freed.

I'll reimplement the logic without pub/sub, but I won't have time until at earliest Sunday, if not later.

**Comment by danielytics** [comment]
I'm looking into a fix for this now.

**Comment by jwr** [comment]
@danielytics it might be worth commenting on the ASYNC-90 issue in JIRA so that everyone knows that people do run into this problem. I voted the issue up, but I don't want to comment, because I don't fully understand the problem.

**Comment by danielcompton** [comment]
This is a beautiful sight. Before and after fixing the issue (at lower and higher resolution):

![screenshot of yourkit java profiler 2015 build 15074 16-10-15 9 14 37 am](https://cloud.githubusercontent.com/assets/811954/10526287/61ccf658-73e6-11e5-9d10-c08630370a62.png)

![screenshot of yourkit java profiler 2015 build 15074 16-10-15 9 20 12 am](https://cloud.githubusercontent.com/assets/811954/10526406/21df0f8a-73e7-11e5-8a6e-bca43f2486e3.png)


https://github.com/apa512/clj-rethinkdb/commit/a274a04a9979d049753e98e07c576d70982072b0

**Comment by danielcompton** [comment]
@jwr I'm pretty sure this fixes the memory leak, it did in my testing on a 64 MB VM. Can you install `master` and test it against your use case please?

**Comment by bhurlow** [comment]
bravo!

**Comment by jwr** [comment]
I'm happy to report that the leak is now gone! Thanks everyone!

(Incidentally — how do you "manually" use an unreleased dependency in leiningen? I worked around this by placing my test code within a checked-out copy of clj-rethinkdb, but that wouldn't always work)

**Comment by apa512** [comment]
```lein install``` should work if I'm understanding correctly what you're trying to do, i.e. clone a repo, make some changes and have those picked up by another project.

**Comment by danielcompton** [comment]
Yep, when you install, it will install a JAR to your local Maven repo. You should see the version come up when you install it. I'll publish a snapshot soon though.

**Comment by jwr** [comment]
Thanks for the tips!

As for the bug fix, I would humbly suggest that it merits an immediate release of a new version.

**Comment by danielytics** [comment]
:+1: agree with @jwr 

Great fix, @danielcompton! Nice and simple. I tried to refactor the code to just use a `mult` directly and not use a `pub` at all, but, while the change was quite small, it caused strange errors and I didn't have time to figure out why. So I'm happy to go with your version.

**Comment by apa512** [comment]
Deployed as ```[com.apa512/rethinkdb "0.11.0"]```.

**Comment by bhurlow** [comment]
:icecream:
